Allow for disabling building of interaction regions

mwyrzykowski opened the pull request:
#### 8622054d0540eb32ccc440c1c97e2bb22a55be6c
<pre>
Allow for disabling building of interaction regions
<a href="https://bugs.webkit.org/show_bug.cgi?id=243128">https://bugs.webkit.org/show_bug.cgi?id=243128</a>

Reviewed by Tim Horton.

Allow for disabling building of interaction regions.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::setInteractionRegionsEnabled):

* Source/WebCore/page/Page.h:
Add function to disable building interaction regions.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _setIsLinkPreview:]):
Disable building interaction regions for link preview web views,
it is not possible to interact with them.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setInteractionRegionsEnabled):
Send message to the webpage.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setInteractionRegionsEnabled):
Forward call to Page::setInteractionRegionsEnabled

* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Add new message to enable / disable building interaction regions.

Canonical link: <a href="https://commits.webkit.org/253307@main">https://commits.webkit.org/253307@main</a>
</pre>

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
```diff
@@ -182,6 +182,10 @@ enum class TapHandlingResult : uint8_t;
 #endif
 
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+@property (nonatomic, readonly) BOOL _isLinkPreview;
```
Should this be in `WKWebViewPrivate.h`?

hortont424 replied:
Hmm, I wonder if this should be a bit that's about the thing you're disabling, not a bit about "is a link preview" that happens to implicitly disable it. I think that's how we usually do this. (I can imagine that there would be other uses for that, also...)

mwyrzykowski replied:
I ended up not needing any additional state variables so I got rid of this

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
```diff
@@ -182,6 +182,10 @@ enum class TapHandlingResult : uint8_t;
 #endif
 
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+@property (nonatomic, readonly) BOOL _isLinkPreview;
```
`setter=_setIsLinkPreview:`?

pxlcoder commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -47,6 +48,10 @@ @interface WKInteractionRegion : NSObject
 @end
 
+@interface UIView (IPI_RemoteLayerTreeInteractionRegionLayers)
+- (UIScrollView *)_scroller;
```
Is there a path to make this SPI / eventually remove our use of IPI?

mwyrzykowski replied:
Not yet but it's also defined in `pal/spi/iOS/UIKitSPI.h` - maybe I should use that instead?

pxlcoder replied:
Looks like it's in `WebKit`, not `PAL` – but yeah let's just import "UIKitSPI.h".

pxlcoder commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -98,6 +112,10 @@ void updateLayersForInteractionRegions(CALayer *layer, const RemoteLayerTreeTran
 
     bool applyBackgroundColorForDebugging = [[NSUserDefaults standardUserDefaults] boolForKey:@"WKInteractionRegionDebugFill"];
 
+    const auto *node = RemoteLayerTreeNode::forCALayer(layer);
```
`auto* node` since `RemoteLayerTreeNode` is a C++ object.

pxlcoder commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -81,6 +86,15 @@ void appendInteractionRegionLayersForLayer(NSMutableArray *sublayers, CALayer *l
+    if (![webView isKindOfClass:[WKWebView class]])
+        return NO;
+
+    return webView._isLinkPreview;
```
This can be written as:

```
if (auto webView = dynamic_objc_cast<WKWebView>([view _scroller].superview))
    return webView._isLinkPreview;

return NO;
```

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -98,6 +108,10 @@ void updateLayersForInteractionRegions(CALayer *layer, const RemoteLayerTreeTran
     bool applyBackgroundColorForDebugging = [[NSUserDefaults standardUserDefaults] boolForKey:@"WKInteractionRegionDebugFill"];
 
+    auto* node = RemoteLayerTreeNode::forCALayer(layer);
+    if (!node || webViewIsLinkPreview(node->uiView()))
```
This is a very late time to have this check; at this point we've already wasted lots of time constructing the regions. I think it would be better to just have SPI that disables the InteractionRegionsEnabled WebPreference.

mwyrzykowski replied:
Good idea, I’ll do that instead

mwyrzykowski commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1660,6 +1660,16 @@ - (BOOL)_isWindowResizingEnabled
 #endif // HAVE(UIKIT_RESIZABLE_WINDOWS)
 
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+- (void)_setIsLinkPreview:(BOOL)isLinkPreview
```
`_setIsLinkPreview` could be renamed now to `_setEnableBuildingInteractionRegions`, does it make sense to make this appear as a general way to toggle interaction regions?

hortont424 replied:
Yes, I was expecting the SPI to get renamed too.

hortont424 commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1660,6 +1660,16 @@ - (BOOL)_isWindowResizingEnabled
+{
+    WKWEBVIEW_RELEASE_LOG("%p -[WKWebView _setIsLinkPreview] page - %p", self, _page.get());
+    if (_page)
+        _page->enableBuildingInteractionRegions(!isLinkPreview);
```
I would expect "setBuildingInteractionRegionsEnabled()" (and I would try to find a way to get "building" out of there -- maybe "setInteractionRegionsEnabled()", or "setShouldBuildInteractionRegions()".

An "enableFoo" taking a bool determining whether it enables or disables is not traditional WebKit style, I think.

hortont424 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1106,7 +1106,15 @@ bool Page::shouldBuildInteractionRegions() const
+void Page::setInteractionRegionsEnabled(bool enable)
+{
+    bool needsUpdate = enable && !shouldBuildInteractionRegions();
+    m_settings->setInteractionRegionsEnabled(enable);
```
This means the Setting and WebPreference get out of sync, right? (and then if something else mutates a WebPreference we'll overwrite it?) I think you could avoid all the new IPC messages *and* keep them in sync if you just mutate the WebPreference (as suggested in https://github.com/WebKit/WebKit/pull/2684#discussion_r928069394).

Also, then you could move the SPI to WKPreferences(Private).h. A lot of the patch dissolves.

But maybe I am missing a good reason it's on the view instead? Maybe it's about WKPreference-sharing?

mwyrzykowski replied:
I'll take another look at this. If I use WKPreferences, I can mutate that value during the lifetime of the web view, for instance when the preview web view is selected, and expect that to work and rebuild the regions? I'm a little unfamiliar with the interaction but I'll take a look

hortont424 replied:
WKPreferences push state down to WebCore::Settings, and update dynamically, yes.

hortont424 replied:
(Via WebKit::WebPreferences)

mwyrzykowski replied:
Oh nice, yes that sounds like a better approach. I’ll do that

mwyrzykowski replied:
@hortont424 and I discussed offline and it seems like this is non-trivial to implement via WKPreferences

hortont424 replied:
(trivial in WebKit, but doesn't work out for clients)

mwyrzykowski commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1692,6 +1692,16 @@ - (BOOL)_isWindowResizingEnabled
+- (void)_setInteractionRegionsEnabled:(BOOL)enabled
+{
+    WKWEBVIEW_RELEASE_LOG("%p -[WKWebView _setInteractionRegionsEnabled] page - %p", self, _page.get());
+    if (_page && _page->preferences().interactionRegionsEnabled())
```
@hortont424 are you ok with this change if it doesn't allow turning on interaction regions when its disabled in the WKPreferences?

So it can get disabled on the WebPage instance but not enabled if interaction regions are disabled in WKPreferences

hortont424 replied:
that logic makes sense to me

mwyrzykowski committed:
Allow for disabling building of interaction regions
https://bugs.webkit.org/show_bug.cgi?id=243128

Reviewed by Tim Horton.

Allow for disabling building of interaction regions.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::setInteractionRegionsEnabled):

* Source/WebCore/page/Page.h:
Add function to disable building interaction regions.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _setIsLinkPreview:]):
Disable building interaction regions for link preview web views,
it is not possible to interact with them.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setInteractionRegionsEnabled):
Send message to the webpage.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setInteractionRegionsEnabled):
Forward call to Page::setInteractionRegionsEnabled

* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Add new message to enable / disable building interaction regions.

Canonical link: https://commits.webkit.org/253307@main
